crates/storage/src/pack.rs

```diff
@@ -53,21 +55,74 @@ use ink_primitives::Key;
 ///
 /// As a general advice pack values together that are frequently used together.
 /// Also pack many very small elements (e.g. `u8`, `bool`, `u16`) together.
-#[derive(Debug, Copy, Clone, scale::Encode, scale::Decode)]
-pub struct Pack<T> {
+#[derive(Debug, Clone)]
+pub struct Pack<T>
+where
+    T: PackedLayout,
+{
     /// The packed `T` value.
     inner: T,
+    /// The key to load the packed value from.
+    ///
+    /// # Note
+    ///
+    /// This can be `None` on contract initialization, but will be
+    /// initialized with a concrete value on `pull_spread`.
+    key: Option<Key>,
+}
+
+impl<T> scale::Encode for Pack<T>
+where
+    T: scale::Encode + PackedLayout,
+{
+    #[inline]
+    fn size_hint(&self) -> usize {
+        <T as scale::Encode>::size_hint(&self.inner)
+    }
+
+    #[inline]
+    fn encode_to<O: scale::Output>(&self, dest: &mut O) {
+        <T as scale::Encode>::encode_to(&self.inner, dest)
+    }
+
+    #[inline]
+    fn encode(&self) -> Vec<u8> {
+        <T as scale::Encode>::encode(&self.inner)
+    }
+
+    #[inline]
+    fn using_encoded<R, F: FnOnce(&[u8]) -> R>(&self, f: F) -> R {
+        <T as scale::Encode>::using_encoded(&self.inner, f)
+    }
 }
 
-impl<T> Pack<T> {
+impl<T> scale::Decode for Pack<T>
+where
+    T: scale::Decode + PackedLayout,
+{
+    fn decode<I: scale::Input>(input: &mut I) -> Result<Self, scale::Error> {
+        Ok(Self::new(<T as scale::Decode>::decode(input)?))
+    }
+}
+
+impl<T> Pack<T>
+where
+    T: PackedLayout,
+{
     /// Creates a new packed value.
     pub fn new(value: T) -> Self {
-        Self { inner: value }
+        Self {
+            inner: value,
+            key: None,
+        }
     }
 
-    /// Returns the packed value.
-    pub fn into_inner(pack: Self) -> T {
```

Comment:
I removed this function, since it is only used in testing and keeping it would require adding `T: Copy`, since `move occurs because pack.inner has type T, which does not implement the Copy trait`.

I mean we could add the trait only for tests, but it will be quite some overhead code just to support three test functions, which can also access the field directly.

Reply:
ah okay I see! :+1: